controllers/k8ssandra/datacenters_test.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	analyticsWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				AnalyticsEnabled: true,
+			},
+		},
+	}
+
+	searchWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				SearchEnabled: true,
+			},
+		},
+	}
+
+	graphWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				GraphEnabled: true,
+			},
+		},
+	}
+
+	mixedWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				GraphEnabled:     true,
+				AnalyticsEnabled: true,
+			},
+		},
+	}
+
+	cassandraWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				GraphEnabled:     false,
+				AnalyticsEnabled: false,
+				SearchEnabled:    false,
+			},
+		},
+	}
+)
+
+func datacentersTest(t *testing.T, ctx context.Context, f *framework.Framework, namespace string) {
+	t.Run("DcUpgradePriorityTest", dcUpgradePriorityTest)
+	t.Run("SortDatacentersForUpgradeTest", sortDatacentersForUpgradeTest)
+	t.Run("SortNoChangeTest", sortNoChangeTest)
+}
+
+func dcUpgradePriorityTest(t *testing.T) {
+	assert := assert.New(t)
+
+	assert.Equal(3, dcUpgradePriority(analyticsWorkloadDc))
+	assert.Equal(1, dcUpgradePriority(searchWorkloadDc))
+	assert.Equal(2, dcUpgradePriority(graphWorkloadDc))
+	assert.Equal(3, dcUpgradePriority(mixedWorkloadDc))
+	assert.Equal(2, dcUpgradePriority(cassandraWorkloadDc))
+}
+
+func sortDatacentersForUpgradeTest(t *testing.T) {
+	assert := assert.New(t)
+
+	datacenters := []api.CassandraDatacenterTemplate{}
+	datacenters = append(datacenters, cassandraWorkloadDc)
+	datacenters = append(datacenters, analyticsWorkloadDc)
+	datacenters = append(datacenters, searchWorkloadDc)
+	datacenters = append(datacenters, graphWorkloadDc)
+	datacenters = append(datacenters, mixedWorkloadDc)
+	sortedDatacenters := sortDatacentersByPriority(datacenters)
+	assert.Equal(5, len(sortedDatacenters), "Expected 5 datacenters")
+	// The analytics enabled DCs should be first, which includes the mixed workload DC
+	assert.True(sortedDatacenters[0].DseWorkloads.AnalyticsEnabled, "Analytics workload DCs should be first")
+	assert.True(sortedDatacenters[1].DseWorkloads.AnalyticsEnabled, "Analytics workload DCs should be first")
+	// Second and third datacenters should be Cassandra and Graph only
+	assert.False(sortedDatacenters[2].DseWorkloads.AnalyticsEnabled || sortedDatacenters[2].DseWorkloads.SearchEnabled, "Analytics workload DC should be first and search should be last")
```

Comment:
The assertions are a bit weird. I have a suggestion: instead of testing the values of these boolean fields, why don't you given distinctive names to each dc, then test the names?